rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,436 @@
+    - These tags can contain only information about data layout, not
+      extra metadata (e.g. zero-point related)
+	- it does not scale well when the number of layouts increases.
+- through memory descriptors. This can be more tedious to use as it is
+  an opaque structure that needs to be queried for
+  information. However, it is very flexible as it allows to express
+  non trivial layout descritions, and can be extended with extra
+  information/meta-data.
+
+
+Layout tags here will have to be different than what is used for the
+oneDNN primitive API, as they would mostly map to part of the
+innerblock component of existing memory::desc objects.
+
+The recommendation here would be to go with layout tags only if no
+handling of metadata and extra information is expected from oneDNN
+block level APIs.
+
+#### Query mecanism for layout
+
+Another question is how to inform user about which layout to use for a
+given brgemm problem.  Similar to oneDNN primitive API, we will expose
+methods to query what is the best layout.  Once again, two options are
+possible:
+- the user create a brgemm object, and they query which layout should
+  be used,
+- we expose a static function that the user can use independantly of
+  the brgemm object.
+
+The first option is optimal in the sense that the dispatched
+implementation has all the knowledge necessary to make the layout
+determination. However, it requires the user to pass a brgemm object
+or a layout tag around when the data pre-packing logic is separate to
+the operator implementation.
+
+The second option could be simpler only if the information necessary to
+make layout determination is readily available and does not require to
+pass extra information around (otherwise it would be equivalent to
+pass layout tag / brgemm object around).  Current internal
+implementation of layout selection depends on highest available ISA,
+input datatype, fp_math mode and accumulation datatype.
+
+In general, we would recommend to tie the layout query to a brgemm
+object for two reasons:
+- from oneDNN implementation perspective, we can extend the
+  information used to select best layout
+- in practice, when data is pre-packed, blocking information (so brgemm
+  object creation parameters) have to be known in order to layout data
+  in an optimal way.
+
+In order to make the case where brgemm objects are recreated for the
+sole purpose of packing data, brgemm obect creation will be split in
+two steps (brgemm::desc and brgemm classes) to avoid jitting
+overheads.  This will introduce some overhead for brgemm objects
+creation though (potential extra copies upon brgemm constructor call).
+
+#### Other considerations
+
+The recommendation as a starting point is to make query mandatory: no
+layout forcing on user side. This simplifies the configurations
+supported by brgemm. Given that we have no support internally for
+transa/tranb, those will not be exposed as well. We will also support
```

Comment:
Do you mean no support for transa/transb on the layout conversion API or on the BRGEMM API? It would be beneficial to have transa/transb (or only transa when b is VNNI) supported at the BRGEMM API.

Reply:
The conversion API will have support for transpose, but not the BRGEMM API.
The main reason is that there is no support for fused transposition in current implementation. We could use that information to generate different kernels internally but in practice, it will likely not be faster than doing the transposition in a temporary buffer before calling brgemm.

Reply:
> The conversion API will have support for transpose, but not the BRGEMM API. The main reason is that there is no support for fused transposition in current implementation. We could use that information to generate different kernels internally but in practice, it will likely not be faster than doing the transposition in a temporary buffer before calling brgemm.

Is it still true if we have HW transposition support in the future? Would it be more flexible to have the transposition flags supported at the API level? BTW, in scaled-dot-product-attention, the first bmm has B transposed.

Reply:
Currently we have no support whatsoever in the implementation for fusing transposition, hence why the proposal does not include it.

If there is value in adding this feature in the future, we would extend the API to support that.